### PR TITLE
Add set and get for Table position

### PR DESCRIPTION
**Pull request by DevinR528**
I'm hoping to switch to tom_edit for [](https://github.com/DevinR528/cargo-sort-ck) the only thing I would need is the ability to reorder the position of tables with segmented headers. Being able to set the position allows me to do this.

Hope this works out! Thanks for a solid toml implementation.

### CODE REVIEW

**Review by ordian — COMMENTED** [review]
Thanks for opening this PR.
One of the goals of this library is to provide a type-safe API for editing TOML documents.
My worry is that with this change, it might be possible to create an invalid TOML file, do you think that's possible?
Also I'd like to see some corner cases tested, e.g. when there is a duplicate position, or a position 0, or usize::max, etc.

File: src/parser/mod.rs [hunk 1]
```diff
@@ -136,7 +135,7 @@ mod tests {
             ("1e6", 1e6),
             ("-2E-2", -2E-2),
             ("6.626e-34", 6.626e-34),
-            ("9_224_617.445_991_228_313", 9_224_617.445_991_228_313),
+            ("9_224_617.445_991_228_313", 9_224_617.445_991_227),
```
[the remark this answers is not in the document]
Comment (DevinR528) [reply src/parser/mod.rs:138]:
```
f64
float has excessive precision
`#[deny(clippy::excessive_precision)]` implied by `#[deny(warnings)]`
for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#excessive_precisionclippyexcessive_precision
```
Although I only get the warning in vscode with rust analyzer running `cargo clippy` doesn't give any errors/warnings?

[the remark this answers is not in the document]
Comment (ordian) [reply src/parser/mod.rs:138]: Hmm, could you merge master to run CI, it should include `cargo clippy` pass?

[the remark this answers is not in the document]
Comment (DevinR528) [reply src/parser/mod.rs:138]: By merge master do you mean my own branch or just push here and run CI?

[the remark this answers is not in the document]
Comment (ordian) [reply src/parser/mod.rs:138]: Thanks, I meant merging origin/master to your branch to include https://github.com/ordian/toml_edit/commit/cd6d1fb2cf43e5472d7eac8fb707bebaf283d9ac. Now it's working.

I think in this case, it's rust-analyzer's fault, because the type here is `f64`, not `f32`, so there's no excessive precision here.

File: tests/test_edit.rs [hunk 2]
```diff
@@ -450,6 +450,67 @@ fn test_sort_values() {
...
+        for (_, table) in root.iter_mut() {
+            as_table!(table).set_position(2)
+        }
+    }).produces_in_original_order(r#"
+        [package]
+        [dependencies]
+        [dev-dependencies]
+        [dependencies.opencl]
+"#);
+}
```
Comment (ordian) [inline tests/test_edit.rs:470]: you're setting every's position to 2, could you alter the test to only do this for one table, e.g. `[dependencies]`?

File: src/table.rs [hunk 3]
```diff
@@ -202,6 +202,19 @@ impl Table {
     pub fn decor(&self, key: &str) -> Option<&Decor> {
         self.items.get(key).map(|kv| &kv.key.decor)
     }
+
+    /// Sets the position of the `Table` within the `Document`.
+    ///
+    /// This will never create invalid toml but, formatting may not be
+    /// preserved (see `tests::test_edit::test_multiple_max_usize_positions`).
```
Comment (ordian) [inline src/table.rs:209]: We should mention this method only affects `Document::to_string_in_original_order()`.

Also see my comment about formatting below.

File: src/table.rs [hunk 4]
```diff
@@ -202,6 +202,19 @@ impl Table {
...
+    pub fn set_position(&mut self, position: usize) {
+        self.position = Some(position);
+    }
+
+    /// The position of the `Table` within the `Document`.
```
Comment (ordian) [inline src/table.rs:214]: It's not clear from a user's perspective, what `None` means.
Though I'm not happy with leaking implementation detail to the public API
```suggestion
    /// The position of the `Table` within the `Document`.
    ///
    /// Returns `None` if the `Table` was created manually (i.e. not via parsing)
    /// in which case its position is set automatically.
```

File: tests/test_edit.rs [hunk 5]
```diff
@@ -450,6 +450,67 @@ fn test_sort_values() {
...
+            as_table!(table).set_position(usize::MAX)
+        }
+    }).produces_in_original_order(r#"        [dependencies.opencl]
+        a=""
+
+        [package]
+        [dependencies]
+        [dev-dependencies]
+"#);
+}
```
Comment (ordian) [inline tests/test_edit.rs:512]: The formatting in the mentioned docs is actually expected, because we treat spaces and comment to the left/right/above a table's header as table's decoration. 

The ordering of tables appears wrong because you only set position for direct children of root (all tables except `dependencies.opencl`.

I think it would be better to be explicit about which tables have their position altered.

**Review by ordian — COMMENTED** [review]
Thanks for adding the tests. There are a few small things to be changed before we can merge this.

File: src/parser/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ordian) [reply src/parser/mod.rs:138]: Could you please revert this change?

File: tests/test_edit.rs (hunk 2 above)
Comment (DevinR528) [reply tests/test_edit.rs:470]: I changed it as you suggested I also set the position of the nested table.

File: src/parser/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DevinR528) [reply src/parser/mod.rs:138]: Ahh, it's because clippy doesn't run for tests unless you `--all-targets`. I did change it back to longer decimal, do you want me to add the `--all-targets` to CI or just leave it?

[the remark this answers is not in the document]
Comment (ordian) [reply src/parser/mod.rs:138]: I see, thanks for digging into this. Yes, please add it here
https://github.com/ordian/toml_edit/blob/cd6d1fb2cf43e5472d7eac8fb707bebaf283d9ac/.github/workflows/ci.yml#L72

**Comment by DevinR528** [comment]
Thanks for adding this feature I appreciate it!!
